### PR TITLE
RandomForest: Support fit multiple models in a single pass

### PR DESCRIPTION
**Pull request by wbo4958**
This PR overrides fitMultiple for RandomForest to fit multiple models in a single pass.

### CODE REVIEW

**Comment by leewyang** [comment]
build

File: python/src/spark_rapids_ml/core.py [hunk 1]
```diff
@@ -484,6 +508,53 @@ def _fit_array_order(self) -> _ArrayOrder:
...
+    """
+    Used by default implementation of Estimator.fitMultiple to produce models in a thread safe
+    iterator. This class handles the gpu versioned of fitMultiple where all param maps should be
+    fit in a single pass.
+
+    Parameters
+    ----------
+    fitMultipleModels : function
+        Callable[[], Transformer] which fits an estimator to a dataset.
+        `fitSingleModel` may be called up to `numModels` times, with a unique index each time.
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply python/src/spark_rapids_ml/core.py:521]: Oops, I copied it from pyspark. I will change it.

File: python/src/spark_rapids_ml/core.py [hunk 2]
```diff
@@ -101,9 +105,14 @@
...
-    "ParamAlias", ("cuml_init", "handle", "num_cols", "part_sizes", "loop")
+    "ParamAlias",
+    ("cuml_init", "handle", "num_cols", "part_sizes", "loop", "fit_multiple_params"),
 )
-param_alias = ParamAlias("cuml_init", "handle", "num_cols", "part_sizes", "loop")
+param_alias = ParamAlias(
+    "cuml_init", "handle", "num_cols", "part_sizes", "loop", "fit_multiple_params"
+)
+
+cuM = TypeVar("cuM", bound="_CumlModel")
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply python/src/spark_rapids_ml/core.py:115]: Done

File: python/src/spark_rapids_ml/core.py [hunk 3]
```diff
@@ -504,17 +575,75 @@ def _create_pyspark_model(self, result: Row) -> "_CumlModel":
...
 
-    def _fit(self, dataset: DataFrame) -> "_CumlModel":
+    def _enable_fit_mulitple_in_single_pass(self) -> bool:
+        """flag to indicate if fitMultiple in a single pass is supported.
+        If not, fallback to super().fitMultiple"""
+        return False
+
+    def fitMultiple(
+        self, dataset: DataFrame, paramMaps: Sequence["ParamMap"]
+    ) -> Iterator[Tuple[int, "_CumlModel"]]:
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply python/src/spark_rapids_ml/core.py:585]: Yes, good suggestion. But I'd like to have that in the following PR to add bound for all the subclass

File: python/src/spark_rapids_ml/tree.py [hunk 4]
```diff
@@ -304,70 +320,78 @@ def _rf_fit(
...
+                fit_multiple_params.append({})
 
-            # Fit a random forest model on the dataset (X, y)
-            rf.fit(X, y, convert_dtype=False)
+            models = []
+            for i in range(len(fit_multiple_params)):
+                tmp_rf_params = rf_params.copy()
+                tmp_rf_params.update(fit_multiple_params[i])
+                rf = cuRf(
+                    n_estimators=n_estimators_per_worker[part_id],
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply python/src/spark_rapids_ml/tree.py:382]: Good catch. I will fix it.

File: python/src/spark_rapids_ml/core.py [hunk 5]
```diff
@@ -504,17 +575,75 @@ def _create_pyspark_model(self, result: Row) -> "_CumlModel":
         """
         raise NotImplementedError()
 
-    def _fit(self, dataset: DataFrame) -> "_CumlModel":
+    def _enable_fit_mulitple_in_single_pass(self) -> bool:
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply python/src/spark_rapids_ml/core.py:578]: Subclasses will decide if enabling fitMultiple on their own. For example, currently, only random forest overrides this function to return True, others (haven't implemented fitMultiple) default to False.

File: python/tests/test_random_forest.py [hunk 6]
```diff
@@ -671,3 +671,46 @@ def test_random_forest_regressor_spark_compat(
...
+
+        assert label_col is not None
+        rf = RFEstimator(bootstrap=False, max_features=1.0, random_state=1.0)
+        rf.setFeaturesCol(features_col)
+        rf.setLabelCol(label_col)
+
+        initial_rf = rf.copy()
+
+        param_maps = [
+            {rf.maxDepth: 3, rf.maxBins: 16},
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply python/tests/test_random_forest.py]: Done.

**Comment by wbo4958** [comment]
build

File: python/src/spark_rapids_ml/core.py [hunk 7]
```diff
@@ -484,6 +513,52 @@ def _fit_array_order(self) -> _ArrayOrder:
         return "F"
 
 
+class _FitMultipleIterator(Generic[CumlModel]):
+    """
+    Used by default implementation of Estimator.fitMultiple to produce models in a thread safe
+    iterator. This class handles the gpu versioned of fitMultiple where all param maps should be
```
Comment (leewyang) [inline python/src/spark_rapids_ml/core.py:519]: "gpu version"?

File: python/src/spark_rapids_ml/core.py [hunk 8]
```diff
@@ -484,6 +513,52 @@ def _fit_array_order(self) -> _ArrayOrder:
...
+
+    Notes
+    -----
+    See :py:meth:`Estimator.fitMultiple` for more info.
+    """
+
+    def __init__(
+        self, fitMultipleModels: Callable[[], List[CumlModel]], numModels: int
+    ):
+        """ """
```
Comment (leewyang) [inline python/src/spark_rapids_ml/core.py:537]: Is this empty docstring intentional?

**Review by leewyang — COMMENTED** [review]
Minor comments, otherwise LGTM

File: python/tests/test_random_forest.py [hunk 9]
```diff
@@ -700,17 +700,32 @@ def test_fit_multiple_in_single_pass(
         initial_rf = rf.copy()
 
         param_maps = [
-            {rf.maxDepth: 3, rf.maxBins: 16},
-            {rf.maxDepth: 3, rf.maxBins: 42},
-            {rf.maxDepth: 6, rf.maxBins: 16},
-            {rf.maxDepth: 6, rf.maxBins: 42},
+            {rf.maxDepth: 3, rf.maxBins: 3, rf.numTrees: 5},
```
Comment (eordentlich) [inline python/tests/test_random_forest.py:703]: Would be good to vary featuresSubsetStrategy, impurity, etc.   All params.

File: python/src/spark_rapids_ml/core.py (hunk 8 above)
Comment (wbo4958) [reply python/src/spark_rapids_ml/core.py:537]: Oops, I didn't realize it. I copied it from pyspark.  Let me remove the empty docstring.

File: python/tests/test_random_forest.py (hunk 9 above)
Comment (wbo4958) [reply python/tests/test_random_forest.py]: Previously, I thought part of the parameters were enough for testing. But this morning, I tried adding featuresSubsetStrategy and found an issue that the fitMultiple or fit didn't update the value that should be mapped.

Really good suggestions.

**Comment by wbo4958** [comment]
build

**Comment by wbo4958** [comment]
Hi @eordentlich @leewyang, when you get a chance, please help to review it again. Thx

File: python/src/spark_rapids_ml/core.py (hunk 7 above)
Comment (wbo4958) [reply python/src/spark_rapids_ml/core.py]: done

**Review by leewyang — COMMENTED** [review]
Minor comments/questions, but otherwise LGTM

**Comment by wbo4958** [comment]
build

File: python/src/spark_rapids_ml/tree.py [hunk 10]
```diff
@@ -304,70 +380,38 @@ def _rf_fit(
...
+                tmp_rf_params.update(fit_multiple_params[i])
+                tmp_rf_params.pop("n_estimators")
+
+                if tmp_rf_params["max_features"] == "auto":
+                    if total_trees[i] == 1:
+                        tmp_rf_params["max_features"] = 1.0
+                    else:
+                        tmp_rf_params["max_features"] = (
+                            "sqrt" if is_classification else (1 / 3.0)
+                        )
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply python/src/spark_rapids_ml/tree.py:392]: good catch, my bad. fixed. Thx

File: python/src/spark_rapids_ml/core.py (hunk 8 above)
Comment (wbo4958) [reply python/src/spark_rapids_ml/core.py]: Sry, new commit has removed it.

**Review by leewyang — APPROVED** [review]

**Pull request merged**